### ISSUE TITLE
Creating a flowchart for visualizing tedana

### ISSUE THREAD

**Issue by emdupre** [ME-ICA/tedana#355]
.@handwerkerd has created a flowchart for visualizing the tedana workflow using graphviz :

[Tedana_ICAandComparisonMetrics_Flowchart.pdf](https://github.com/ME-ICA/tedana/files/3373607/Tedana_ICAandComparisonMetrics_Flowchart.pdf)

This would replace the illustration we currently have in the documentation, 

![](https://tedana.readthedocs.io/en/latest/_images/tedana-workflow.png)

It'd be great to discuss the aims and implementation of this visualization !

**Comment by emdupre** [comment]
Some initial feedback I have -- obviously there's quite a lot going on right now, but that's definitely useful for figuring out all of the pieces of the workflow ! The question to me, then, is what is the end goal.

In Nipype, graphviz is used to create graphs of the exact processing strategy a user employed. Do we want to do something similar here ? Personally I'm not sure we have enough options to warrant that, but would love to hear other folks's opinions. 

Alternatively, if this is just for documentation, does it make sense to do this in graphviz ? Or should we focus on creating something interactive / illustrative rather than programmatically generated.

**Comment by emdupre** [comment]
Copying in other feedback from gitter:

Joshua Teves @jbteves 10:54
>  I agree with the messy flowchart guiding further modularization; in an ideal world our code would make a pretty flowchart

Taylor Salo @tsalo 11:05
> @handwerkerd I think the summary metrics and metric maps (e.g., F_S0) could go in a table instead of the flowchart. Then the flowchart could maybe contain the common processes underlying the metrics (e.g., “cluster” vs “non-cluster” and “voxel-wise fit statistics”), which would cut the number of boxes at the bottom of the flowchart roughly in half, because T2 and S0-specific metrics wouldn’t be duplicated.

**Comment by handwerkerd** [comment]
The very rough draft version of the flowchart & GraphViz code are attached.  Just to copy key points from the Gitter conversation where this started: 
The attached version is just the pipeline through all the metrics that are outputted into comptable (i.e. how the metrics that’s used in the decision tree are calculated). The followup would be a similar flow chart for the actual decision tree process.
I don't think the current version is useful for end users (too messy), but I think it could guide improved modularization of the code and would be useful for potential developers.
[Tedana_ICAandComparisonMetrics_Flowchart.pdf](https://github.com/ME-ICA/tedana/files/3373620/Tedana_ICAandComparisonMetrics_Flowchart.pdf)

[Tedana_ICAandComparisonMetrics_Flowchart.dot.zip](https://github.com/ME-ICA/tedana/files/3373629/Tedana_ICAandComparisonMetrics_Flowchart.dot.zip)

**Comment by handwerkerd** [comment]
I'll second the idea of a table. The trouble I was having in putting everything in one flow chart is that the first part (what's above) was really about getting to a bunch of metrics which each have one value per component. It was really difficult to figure out what was necessary for each metric to be calculated. I can see the revised version having 3 parts:
1. The current flow chart stopping short of the metrics box. There are only 5 volumes that feed into all the metrics that could then be clearly highlighted.
2. A table with the metrics that includes clear info on which of the 5 volumes are inputs. Interestingly the 5 volumes have overlapping information, but only 2 (variance & normalized variance) take identical inputs. This makes me slightly more confident that each of these really does measures something unique & potentially useful.
3. A Decision tree that relies on knowing the names of the metrics listed in the table.

I think it is possible to make links in Graphviz so it would be fairly easy to make a low-level amount of interactivity.

**Comment by handwerkerd** [comment]
To answer @emdupre's question, I think it would greatly strengthen the code to get it to a state where a flow chart can automatically be generated based on what options a user applies. If every metric & decision tree step is modularly calculated and well-commented, adding blocks of code that could generate such a Graphviz flow chart would be very do-able. That said, I see that as low hanging fruit once the code is modular and flexible enough to make that possible rather than a stand-alone goal.

**Comment by handwerkerd** [comment]
One additional note on the current draft is that PCA dimensionality reduction is a single black box. Given there are a few options for this in consideration, I figured it was best to keep that simple, but that's also a point that might need another flowchart or table that includes expanded details.

**Comment by jbteves** [comment]
@handwerkerd I agree with all of the above, with the additional comment that maybe we should actually make two charts: what we have, and what we think would look better. That should help guide us towards where the code should be refactored. WDYT, @emdupre ?

**Comment by handwerkerd** [comment]
FWIW, I think the current flow chart IS what would look better in the code. Much of the headache in putting this together is how the metrics are calculated in a few different places with the precursors to the metrics in other not-always-logical places. A re-org of code where all the metrics for the decision tree were calculated with a coherent structure of functions would be a huge improvement in readability and modularity.

**Comment by monicayao** [comment]
I have a small idea! Since there are so many arrows, a good thing may be to color code the arrows near the bottom half. And also, a great thing (in my opinion) that would encourage more people to contribute and code for tedana is to somehow incorporate the input/output (dimension/type, perhaps) associated with each step. This may make the flowchart more cluttered, but I think would be very helpful!

**Comment by handwerkerd** [comment]
The updated flow chart plus table for the decision tree metrics and a new flow chart for the decision tree itself are attached. 
Note that there are comments in the GraphViz (compressed) dot files that might help explain some of my choices. I also tried to make direct links to code sections for the decision tree flow chart and I plan to do that for the metric flow chart as well.

Both of these can definitely be cleaned up,  it might be possible to add some html links within nodes, and I still haven't made a simplified flow chart for the whole process. Still I figured this was the right time for feedback.

Also, I went through the decision tree, in detail, with @jwe080 and wanted to make sure she gets credit here.
[Tedana_ICAandComparisonMetrics_Flowchart_v2.pdf](https://github.com/ME-ICA/tedana/files/3437322/Tedana_ICAandComparisonMetrics_Flowchart_v2.pdf)
[Tedana_ICADecisionTree_Flowchart.pdf](https://github.com/ME-ICA/tedana/files/3437323/Tedana_ICADecisionTree_Flowchart.pdf)
[Tedana_FlowCharts_v2.tar.gz](https://github.com/ME-ICA/tedana/files/3437328/Tedana_FlowCharts_v2.tar.gz)

**Comment by tsalo** [comment]
Both look great. I have some notes though.
For the metric workflow:
- `kappa ratio` is calculated within the decision tree, so it might be misleading to include it in the metric calculation workflow.
- The metric maps `F_S0_clmaps`, `F_R2_clmaps`, `Br_S0_clmaps`, and `Br_R2_clmaps` aren't included in the tree.
- It would be nice to have a node for the dimensionally reduced data between the PCA and ICA nodes.
- `normvarexp` isn't used by the ICA decision tree, so we can drop that I think.

For the decision tree:
- I think the decision tree figure is still a bit confusing, although I'm not sure it's possible to make it completely clear.
- Can we distinguish between what one might call "absolute" steps and "relative" ones? I002 and I003 are determined on a component-by-component basis, while the rest rely on elbows and patterns _across_ components to make decisions. I004 and I005 also use variance explained, which isn't described in the flowchart.
- The three highest variance components are retained in `unclf` even if they're dropped from `ncls`, so they may end up being accepted later. I'm also not sure if selecting the three highest variance remaining components is what that step is _supposed_ to do. There are certainly easier ways of doing that than a for loop with the whole "diff" thing.
    - If that is what the step should do, then this might be better text: "Does the component have one of the top three variance explained values of the remaining components?". That way it's more of a decision step with a Yes and a No.
- The arrow with "No All remaining components regardless of provisional acceptance" is a little confusing. Perhaps you could put an oval in the middle with "Unclassified components" to make it clear that the two No arrows operate in parallel on overlapping, but different, subsets of the data?
- There's an arrow missing the text "yes" in the middle of the figure.

**Comment by handwerkerd** [comment]
Replies to some of @tsalo's comment.

- `kappa ratio` and `rho elbow` both involve information not known before the decision tree, but I can't think of a way to include them there without making a mess of the decision tree. I decided to include them in the the metrics flow chart with an *. Other ideas are welcome.

- `F_S0_clmaps`, `F_R2_clmaps`, `Br_S0_clmaps`, and `Br_R2_clmaps` are all intermediate calculations that are used for the metrics, but aren't used elsewhere. They're essentially maps in the chart that are cluster thresholded various ways. I skipped naming them because I thought they didn't add much, but I didn't include `Z_clMap`. This is a subjective decision that could use more thought.

- I decided to put "dimensionality reduction" in the PCA node. A previous version had it separate & it just seemed to add clutter, but this also can be revisited.

- `normvarexp` should be explained because it's an output in comp_table, but it's not in the decision tree. We should identify any other relevant outputs and make sure they're included in the flow chart even if they aren't in the decision tree.


Your questions on the decision tree are great because they mean you follow enough to realize that there are parts that might deserve re-thinking. :)

- I like the idea of formatting (color/shape/fill) based on the type of of decisions in each step, but will need to give some more thought on what makes the most visual sense.

- For the 3 highest variance components, as best as I can tell, they are excluded from the elbow calculations and are then included in everything afterwards. That's why there are two "No" arrows leaving the I002-5 box. That was the best way I could think of to visualize that, but perhaps there's a better way. 

- The "No: Ignore everything" arrow is just that. If there are no provisionally accepted components, the rest of the tree isn't run. Anything that wasn't rejected is ignored and there are no accepted components.  I'll continue to think of a better visualization for that. FWIW, every step is only run on one group of components so there are never parallel processing streams. The only parallel point is with the high variance components returning to the decision tree after a few skipped steps.

- Is the missing "yes" between provisionally accepted components" and "I006" If so that's intentional? There is a class of "provisionally accepted" that is evaluated in I006, but there isn't a decision point between the two.

Everything here assumes I traced the code correctly.

**Comment by emdupre** [comment]
@handwerkerd is there a plan to add your updated flowcharts into the docs somewhere ?

**Comment by tsalo** [comment]
It would also be nice to have a publication-ready top-level figure like [this one](https://raw.githubusercontent.com/oesteban/fmriprep/38a63e9504ab67812b63813c5fe9af882109408e/docs/_static/fmriprep-workflow-all.png), with brains and things.

**Comment by handwerkerd** [comment]
I definitely plan to have a flowchart that can be dynamically generated and output with each decision tree that is run. I've set up the decision tree function outputs to make this practical. In the short term, I don't want to hold up merging the decision tree code until it can also create flow charts, but it is part of the plan. That said, I've figured out hand-written GraphViz, but I know there's a good wrapper to dynamically generate a GraphViz markup in python. Has anyone here done this or knows of a good example in another repository?

**Comment by handwerkerd** [comment]
This is still in progress, with the goal of integrating flowchart creation with the modularized decision tree: #475 & #476

**Comment by tsalo** [comment]
I'm working a bit on the JOSS manuscript and was wondering if anyone had made any progress on the static publication-quality figure (rather than the dynamically-generated workflow graphs we're interested in for the modularized workflow)?

**Comment by handwerkerd** [comment]
The static figures that are linked to [here](https://github.com/ME-ICA/tedana/issues/355#issuecomment-515593755) should still be fairly accurate. You or I can tweak them depending on your specific needs.

**Comment by tsalo** [comment]
I don't think we need that level of detail. I was thinking something more aesthetically pleasing, with a top-level schematic of the workflow, like this:

![image](https://user-images.githubusercontent.com/8228902/100274105-e1017380-2f2b-11eb-9c4d-098d04596f02.png)

**Comment by handwerkerd** [comment]
That's definitely possible, but I don't think something like that currently exists. I'm definitely not getting to it this week, but check back with me in a week or two.

**Comment by tsalo** [comment]
I have just started using OmniGraffle, so I am thinking of working on this a bit for the manuscript. @handwerkerd do you have OmniGraffle, or do you use something else for publication-quality figures?

**Comment by emdupre** [comment]
This figure came up in discussions on #\680. Do we have any updates, here ? Although a Esteban-style figure might be a bit further off, should we at least update the flow-chat itself, e.g. to remove the link between the input data and TEDPCA ?

**Comment by tsalo** [comment]
I ran out of the trial period and haven't bought an OmniGraffle license... so no progress to report 😞

**Comment by notZaki** [comment]
I had made a few simplified flow charts using [whimsical](https://whimsical.com/) when I was initially trying to figure out the code. Perhaps that could be useful for the flow charts? I've attached two examples below, though they might not necessarily be accurate.

![image](https://user-images.githubusercontent.com/9019681/108583751-85b00900-7301-11eb-9eff-c5cd53c9c5c4.png)

![image](https://user-images.githubusercontent.com/9019681/108583761-93fe2500-7301-11eb-94ea-1ae1451a55eb.png)

**Comment by handwerkerd** [comment]
The GraphViz code I've included above still models the current algorithm. The other versions are slightly prettier, but the nice thing about about GraphViz is it should be eventually possible to automatically generate those figures. It wouldn't be hard to move those flow charts into the documentation.

**Comment by emdupre** [comment]
> The other versions are slightly prettier, but the nice thing about about GraphViz is it should be eventually possible to automatically generate those figures. It wouldn't be hard to move those flow charts into the documentation.

I definitely would love to output automatically generated flowcharts ! This could become especially important as we allow users more flexibility in the metric calculation -- and could be bundled alongside the reports !

I like the idea of updating the documentation with one of the whimsical flowcharts too, though (thanks for sharing those, @notZaki !). It would probably make sense to keep a high-level figure in the docs even if the automatically generated flowcharts and more specific to an individual call (much like fMRIPrep has a high-level figure that @tsalo shared above, and then low-level flowcharts for several sub-routines within the workflow). WDYT ?

**Comment by handwerkerd** [comment]
My biased opinion is that the information in the flow chart used in the [tedana OHBM 2020 poster](https://github.com/ME-ICA/tedana-ohbm-2020/blob/master/docs/OHBM2020_1083_tedana_poster_with_links.pdf) is currently the most accurate summary of what the code is doing without being overly complicated. If it's being added in a static form to the documentation, it can be redone in a visually more appealing format. If we want to add more detail to the static documentation, an explanation of what each metric takes as input could also be added. That info is in the ComparisonMetrics Flowchart [from this comment](https://github.com/ME-ICA/tedana/issues/355#issuecomment-515593755). I am happy to help anyone who wants to make visually more appealing versions of these, but I don't think I'll have time to do this myself in the semi-near future.
<img width="212" alt="image" src="https://user-images.githubusercontent.com/7406227/108731844-f79f7280-74fa-11eb-8c00-d13b13e11785.png">
